## 1. Creating Spark Session

In [0]:
from pyspark.sql import SparkSession

spark = SparkSession.builder \
.appName("Data_Transformation_Silver_Layer") \
.getOrCreate()

In [0]:
spark

***
### 2. Connecting to ADLS Storage Account to access datasets

In [0]:
storage_account = "ecommoliststorageaccount"
application_id = "01df8d6d-8848-4f6e-98fa-b7b6b9fd0195"
directory_id = "605c3950-6e77-4105-a1d5-40ed5d49f86b"

spark.conf.set(f"fs.azure.account.auth.type.{storage_account}.dfs.core.windows.net", "OAuth")
spark.conf.set(f"fs.azure.account.oauth.provider.type.{storage_account}.dfs.core.windows.net", "org.apache.hadoop.fs.azurebfs.oauth2.ClientCredsTokenProvider")
spark.conf.set(f"fs.azure.account.oauth2.client.id.{storage_account}.dfs.core.windows.net", application_id)
spark.conf.set(f"fs.azure.account.oauth2.client.secret.{storage_account}.dfs.core.windows.net", "ror8Q~RIGdkmIHMNYf1j4tX7ubsSOziUtLYJlaT_")
spark.conf.set(f"fs.azure.account.oauth2.client.endpoint.{storage_account}.dfs.core.windows.net", f"https://login.microsoftonline.com/{directory_id}/oauth2/token")

***
### 3. Reading Data into Databricks

In [0]:
storage_account = "ecommoliststorageaccount"

base_path = f"abfss://olistdata@{storage_account}.dfs.core.windows.net/bronze/"

customers_path = base_path + "olist_customers_dataset.csv"
geolocation_path = base_path + "olist_geolocation_dataset.csv"
order_items_path = base_path + "olist_order_items_dataset.csv"
order_payments_path = base_path + "olist_order_payments.csv"
order_reviews_path = base_path + "olist_order_reviews_dataset.csv"
orders_path = base_path + "olist_orders_dataset.csv"
products_path = base_path + "olist_products_dataset.csv"
sellers_path = base_path + "olist_sellers_dataset.csv"


customers_df = spark.read.csv(customers_path, header = True, inferSchema = True)
geolocation_df = spark.read.csv(geolocation_path, header = True, inferSchema = True)
order_items_df = spark.read.csv(order_items_path, header = True, inferSchema = True)
order_payments_df = spark.read.csv(order_payments_path, header = True, inferSchema = True)
order_reviews_df = spark.read.csv(order_reviews_path, header = True, inferSchema = True)
orders_df = spark.read.csv(orders_path, header = True, inferSchema = True)
products_df = spark.read.csv(products_path, header = True, inferSchema = True)
sellers_df = spark.read.csv(sellers_path, header = True, inferSchema = True)


#### Creating function to describe datasets:

In [0]:
def describe_df(df_name, df):
    print(f"DataFrame '{df_name}' has {df.count()} rows and {len(df.columns)} columns.")
    print("-"*50)
    print("Schema:")
    print(df.printSchema())
    print("-"*50)
    print(f"Showing top 10 dataframe: {df_name}'s contents:")
    display(df.limit(10))


In [0]:
describe_df("orders", orders_df)

DataFrame 'orders' has 99441 rows and 8 columns.
--------------------------------------------------
Schema:
root
 |-- order_id: string (nullable = true)
 |-- customer_id: string (nullable = true)
 |-- order_status: string (nullable = true)
 |-- order_purchase_timestamp: timestamp (nullable = true)
 |-- order_approved_at: timestamp (nullable = true)
 |-- order_delivered_carrier_date: timestamp (nullable = true)
 |-- order_delivered_customer_date: timestamp (nullable = true)
 |-- order_estimated_delivery_date: timestamp (nullable = true)

None
--------------------------------------------------
Showing top 10 dataframe: orders's contents:


order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02T10:56:33Z,2017-10-02T11:07:15Z,2017-10-04T19:55:00Z,2017-10-10T21:25:13Z,2017-10-18T00:00:00Z
53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24T20:41:37Z,2018-07-26T03:24:27Z,2018-07-26T14:31:00Z,2018-08-07T15:27:45Z,2018-08-13T00:00:00Z
47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08T08:38:49Z,2018-08-08T08:55:23Z,2018-08-08T13:50:00Z,2018-08-17T18:06:29Z,2018-09-04T00:00:00Z
949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18T19:28:06Z,2017-11-18T19:45:59Z,2017-11-22T13:39:59Z,2017-12-02T00:28:42Z,2017-12-15T00:00:00Z
ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13T21:18:39Z,2018-02-13T22:20:29Z,2018-02-14T19:46:34Z,2018-02-16T18:17:02Z,2018-02-26T00:00:00Z
a4591c265e18cb1dcee52889e2d8acc3,503740e9ca751ccdda7ba28e9ab8f608,delivered,2017-07-09T21:57:05Z,2017-07-09T22:10:13Z,2017-07-11T14:58:04Z,2017-07-26T10:57:55Z,2017-08-01T00:00:00Z
136cce7faa42fdb2cefd53fdc79a6098,ed0271e0b7da060a393796590e7b737a,invoiced,2017-04-11T12:22:08Z,2017-04-13T13:25:17Z,null,null,2017-05-09T00:00:00Z
6514b8ad8028c9f2cc2374ded245783f,9bdf08b4b3b52b5526ff42d37d47f222,delivered,2017-05-16T13:10:30Z,2017-05-16T13:22:11Z,2017-05-22T10:07:46Z,2017-05-26T12:55:51Z,2017-06-07T00:00:00Z
76c6e866289321a7c93b82b54852dc33,f54a9f0e6b351c431402b8461ea51999,delivered,2017-01-23T18:29:09Z,2017-01-25T02:50:47Z,2017-01-26T14:16:31Z,2017-02-02T14:08:10Z,2017-03-06T00:00:00Z
e69bfb5eb88e0ed6a785585b27e16dbf,31ad1d1b63eb9962463f764d4e6e0c9d,delivered,2017-07-29T11:55:02Z,2017-07-29T12:05:32Z,2017-08-10T19:45:24Z,2017-08-16T17:14:30Z,2017-08-23T00:00:00Z


***
### 4. Data Cleansing

***
#### I. Identifying Missing Values:-
##### - Create function to identify missing values from each column in a dataframe

In [0]:
from pyspark.sql.functions import col, when, count

def missing_values(df, df_name):
    print(f"Missing values in {df_name} dataframe:")
    df.select([count(when(col(c).isNull(), 1)).alias(c) for c in df.columns]).show()

In [0]:
missing_values(customers_df, "customers")

Missing values in customers dataframe:
+-----------+------------------+------------------------+-------------+--------------+
|customer_id|customer_unique_id|customer_zip_code_prefix|customer_city|customer_state|
+-----------+------------------+------------------------+-------------+--------------+
|          0|                 0|                       0|            0|             0|
+-----------+------------------+------------------------+-------------+--------------+



In [0]:
missing_values(geolocation_df, "geolocation")

Missing values in geolocation dataframe:
+---------------------------+---------------+---------------+----------------+-----------------+
|geolocation_zip_code_prefix|geolocation_lat|geolocation_lng|geolocation_city|geolocation_state|
+---------------------------+---------------+---------------+----------------+-----------------+
|                          0|              0|              0|               0|                0|
+---------------------------+---------------+---------------+----------------+-----------------+



In [0]:
missing_values(order_items_df, "order_items")

Missing values in order_items dataframe:
+--------+-------------+----------+---------+-------------------+-----+-------------+
|order_id|order_item_id|product_id|seller_id|shipping_limit_date|price|freight_value|
+--------+-------------+----------+---------+-------------------+-----+-------------+
|       0|            0|         0|        0|                  0|    0|            0|
+--------+-------------+----------+---------+-------------------+-----+-------------+



In [0]:
missing_values(order_payments_df, "order_payments")

Missing values in order_payments dataframe:
+--------+------------------+------------+--------------------+-------------+
|order_id|payment_sequential|payment_type|payment_installments|payment_value|
+--------+------------------+------------+--------------------+-------------+
|       0|                 0|           0|                   0|            0|
+--------+------------------+------------+--------------------+-------------+



In [0]:
missing_values(order_reviews_df, "order_reviews")

Missing values in order_reviews dataframe:
+---------+--------+------------+--------------------+----------------------+--------------------+-----------------------+
|review_id|order_id|review_score|review_comment_title|review_comment_message|review_creation_date|review_answer_timestamp|
+---------+--------+------------+--------------------+----------------------+--------------------+-----------------------+
|        1|    2236|        2380|               92157|                 63079|                8764|                   8785|
+---------+--------+------------+--------------------+----------------------+--------------------+-----------------------+



In [0]:
missing_values(orders_df, "orders")

Missing values in orders dataframe:
+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+
|order_id|customer_id|order_status|order_purchase_timestamp|order_approved_at|order_delivered_carrier_date|order_delivered_customer_date|order_estimated_delivery_date|
+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+
|       0|          0|           0|                       0|              160|                        1783|                         2965|                            0|
+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+



In [0]:
missing_values(products_df, "products")

Missing values in products dataframe:
+----------+---------------------+-------------------+--------------------------+------------------+----------------+-----------------+-----------------+----------------+
|product_id|product_category_name|product_name_lenght|product_description_lenght|product_photos_qty|product_weight_g|product_length_cm|product_height_cm|product_width_cm|
+----------+---------------------+-------------------+--------------------------+------------------+----------------+-----------------+-----------------+----------------+
|         0|                  610|                610|                       610|               610|               2|                2|                2|               2|
+----------+---------------------+-------------------+--------------------------+------------------+----------------+-----------------+-----------------+----------------+



In [0]:
missing_values(sellers_df, "sellers")

Missing values in sellers dataframe:
+---------+----------------------+-----------+------------+
|seller_id|seller_zip_code_prefix|seller_city|seller_state|
+---------+----------------------+-----------+------------+
|        0|                     0|          0|           0|
+---------+----------------------+-----------+------------+



**> Missing values identified in orders, products & reviews dataframe. We will need to handle these missing values in this data cleaning stage.**

***
#### II. Handling Missing Values:-

##### - Dropping missing values (for critical columns like identifiers, e.g., review_id, etc.)
##### - Filling missing values (for continuous datatype columns)
##### - Impute missing values (for numerical datatype columns)

 ***
 **i. Dropping missing values (for critical columns)**

In [0]:
missing_values(order_reviews_df, "order_reviews")

Missing values in order_reviews dataframe:
+---------+--------+------------+--------------------+----------------------+--------------------+-----------------------+
|review_id|order_id|review_score|review_comment_title|review_comment_message|review_creation_date|review_answer_timestamp|
+---------+--------+------------+--------------------+----------------------+--------------------+-----------------------+
|        1|    2236|        2380|               92157|                 63079|                8764|                   8785|
+---------+--------+------------+--------------------+----------------------+--------------------+-----------------------+



In [0]:
from functools import reduce
from pyspark.sql import DataFrame, functions as F


def clean_dataframe(df: DataFrame, df_name: str, *cols: str) -> DataFrame:
    """
    Cleans a PySpark DataFrame.

    Functionality:
    1. If no columns are specified:
       - Drops rows where ALL columns contain NULL values.

    2. If specific columns are specified:
       - Drops rows where ANY specified column contains a NULL value.

    Args:
        df (DataFrame): Input PySpark DataFrame.
        df_name (str): Name of the DataFrame.
        *cols (str): Optional columns to check for NULL values.

    Returns:
        DataFrame: Cleaned DataFrame.
    """

    print(f"Cleaning {df_name}...")

    # ---------------------------------------------------------
    # CASE 1: No columns are specified
    # Drop rows only when ALL columns are NULL
    # ---------------------------------------------------------
    if not cols:

        # Create a condition:
        # column_1 IS NULL AND column_2 IS NULL AND ...
        all_null_condition = reduce(
            lambda x, y: x & y,
            [F.col(column).isNull() for column in df.columns]
        )

        # Identify rows where all columns are NULL
        rows_to_drop = df.filter(all_null_condition)

        # Count the rows before dropping
        null_row_count = rows_to_drop.count()

        if null_row_count > 0:
            print(
                f"Warning: {df_name} has {null_row_count} row(s) "
                "where all columns contain NULL values."
            )

            # Show the rows before dropping them
            rows_to_drop.show()

            print("-" * 50)
            print("Dropping these rows...")

            # Keep rows that are NOT completely NULL
            df = df.filter(~all_null_condition)

            print("The rows have been dropped.")
            print("-" * 50)

        else:
            print(f"No rows with all NULL values found in {df_name}.")

    # ---------------------------------------------------------
    # CASE 2: Specific columns are specified
    # Drop rows when ANY specified column is NULL
    # ---------------------------------------------------------
    else:

        # Validate that all specified columns exist in the DataFrame
        invalid_columns = [
            column for column in cols
            if column not in df.columns
        ]

        if invalid_columns:
            raise ValueError(
                f"These columns do not exist in {df_name}: "
                f"{invalid_columns}"
            )

        # Create a condition:
        # column_1 IS NULL OR column_2 IS NULL OR ...
        any_null_condition = reduce(
            lambda x, y: x | y,
            [F.col(column).isNull() for column in cols]
        )

        # Identify rows that have NULL in any specified column
        rows_to_drop = df.filter(any_null_condition)

        # Count the rows before dropping
        null_row_count = rows_to_drop.count()

        if null_row_count > 0:
            print(
                f"Warning: {df_name} has {null_row_count} row(s) "
                "with NULL values in one or more of the specified columns."
            )

            # Show the rows before dropping them
            rows_to_drop.show()

            print("-" * 50)
            print("Dropping these rows...")

            # Keep only rows where ALL specified columns are NOT NULL
            df = df.filter(~any_null_condition)

            print(
                "The rows with NULL values in any of the mentioned "
                "columns have been dropped."
            )
            print("-" * 50)

        else:
            print(
                f"No NULL values found in the specified columns "
                f"of {df_name}."
            )

    return df

In [0]:
# Drop rows where all columns are NULL

cleaned_order_reviews_df = clean_dataframe(order_reviews_df, "order_reviews")

Cleaning order_reviews...
No rows with all NULL values found in order_reviews.


In [0]:
missing_values(cleaned_order_reviews_df, "order_reviews")

Missing values in order_reviews dataframe:
+---------+--------+------------+--------------------+----------------------+--------------------+-----------------------+
|review_id|order_id|review_score|review_comment_title|review_comment_message|review_creation_date|review_answer_timestamp|
+---------+--------+------------+--------------------+----------------------+--------------------+-----------------------+
|        1|    2236|        2380|               92157|                 63079|                8764|                   8785|
+---------+--------+------------+--------------------+----------------------+--------------------+-----------------------+



In [0]:
cleaned_order_reviews_df.filter(col("review_id").isNull()).show()

+---------+--------------------+--------------------+--------------------+----------------------+--------------------+-----------------------+
|review_id|            order_id|        review_score|review_comment_title|review_comment_message|review_creation_date|review_answer_timestamp|
+---------+--------------------+--------------------+--------------------+----------------------+--------------------+-----------------------+
|     NULL|material de boa q...|pensei que era ma...|mas e muito bonit...|   2018-01-06 00:00:00| 2018-01-08 14:20:31|                   NULL|
+---------+--------------------+--------------------+--------------------+----------------------+--------------------+-----------------------+



In [0]:
cleaned_order_reviews_df.filter(col("review_id").isNull() | col("order_id").isNull()).show()

+--------------------+--------+------------+--------------------+----------------------+--------------------+-----------------------+
|           review_id|order_id|review_score|review_comment_title|review_comment_message|review_creation_date|review_answer_timestamp|
+--------------------+--------+------------+--------------------+----------------------+--------------------+-----------------------+
|,2018-02-16 00:00...|    NULL|        NULL|                NULL|                  NULL|                NULL|                   NULL|
|A entrega foi efe...|    NULL|        NULL|                NULL|                  NULL|                NULL|                   NULL|
|O produto já come...|    NULL|        NULL|                NULL|                  NULL|                NULL|                   NULL|
|    Estou satisfeita|    NULL|        NULL|                NULL|                  NULL|                NULL|                   NULL|
|,2017-06-02 00:00...|    NULL|        NULL|                NU

In [0]:
# Drop rows where the specified critical columns values are NULL

cleaned_order_reviews_df = clean_dataframe(order_reviews_df, "order_review", "review_id", "order_id")

Cleaning order_review...
+--------------------+--------+------------+--------------------+----------------------+--------------------+-----------------------+
|           review_id|order_id|review_score|review_comment_title|review_comment_message|review_creation_date|review_answer_timestamp|
+--------------------+--------+------------+--------------------+----------------------+--------------------+-----------------------+
|,2018-02-16 00:00...|    NULL|        NULL|                NULL|                  NULL|                NULL|                   NULL|
|A entrega foi efe...|    NULL|        NULL|                NULL|                  NULL|                NULL|                   NULL|
|O produto já come...|    NULL|        NULL|                NULL|                  NULL|                NULL|                   NULL|
|    Estou satisfeita|    NULL|        NULL|                NULL|                  NULL|                NULL|                   NULL|
|,2017-06-02 00:00...|    NULL|      

In [0]:
missing_values(cleaned_order_reviews_df, "order_reviews")

Missing values in order_reviews dataframe:
+---------+--------+------------+--------------------+----------------------+--------------------+-----------------------+
|review_id|order_id|review_score|review_comment_title|review_comment_message|review_creation_date|review_answer_timestamp|
+---------+--------+------------+--------------------+----------------------+--------------------+-----------------------+
|        0|       0|         144|               89922|                 60843|                6528|                   6548|
+---------+--------+------------+--------------------+----------------------+--------------------+-----------------------+



***
**ii. Fill missing values (for continuous datatype columns)**

In [0]:
missing_values(cleaned_order_reviews_df, "order_reviews")

Missing values in order_reviews dataframe:
+---------+--------+------------+--------------------+----------------------+--------------------+-----------------------+
|review_id|order_id|review_score|review_comment_title|review_comment_message|review_creation_date|review_answer_timestamp|
+---------+--------+------------+--------------------+----------------------+--------------------+-----------------------+
|        0|       0|         144|               89922|                 60843|                6528|                   6548|
+---------+--------+------------+--------------------+----------------------+--------------------+-----------------------+



In [0]:
cleaned_order_reviews_df.filter(
                col("review_score").isNull() | 
                col("review_comment_title").isNull() |
                col("review_comment_message").isNull() |
                col("review_creation_date").isNull() |
                col("review_answer_timestamp").isNull() 
        ).display()

review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,null,null,2018-01-18 00:00:00,2018-01-18 21:46:59
80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,null,null,2018-03-10 00:00:00,2018-03-11 03:05:13
228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,null,null,2018-02-17 00:00:00,2018-02-18 14:36:24
e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,null,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,null,Parabéns lojas lannister adorei comprar pela Internet seguro e prático Parabéns a todos feliz Páscoa,2018-03-01 00:00:00,2018-03-02 10:26:53
15197aa66ff4d0650b5434f1b46cda19,b18dcdf73be66366873cd26c5724d1dc,1,null,null,2018-04-13 00:00:00,2018-04-16 00:39:37
07f9bee5d1b850860defd761afa7ff16,e48aa0d2dcec3a2e87348811bcfdf22b,5,null,null,2017-07-16 00:00:00,2017-07-18 19:30:34
7c6400515c67679fbee952a7525281ef,c31a859e34e3adac22f376954e19b39d,5,null,null,2018-08-14 00:00:00,2018-08-14 21:36:06
a3f6f7f6f433de0aefbb97da197c554c,9c214ac970e84273583ab523dfafd09b,5,null,null,2017-05-17 00:00:00,2017-05-18 12:05:37
c9cfd2d5ab5911836ababae136c3a10c,cdf9aa68e72324eeb25c7de974696ee2,5,null,null,2017-12-23 00:00:00,2017-12-26 14:36:03


For these columns, I would not use the same default value for every field. The best replacement depends on the column's meaning and data type.

Assuming this is the Olist order_reviews_dataset, I recommend:

| Column                    | Data type | Recommended NULL replacement                                         | Why                                                              |
| ------------------------- | --------- | -------------------------------------------------------------------- | ---------------------------------------------------------------- |
| `review_score`            | Integer   | **Do not fill with 0**; preferably keep NULL or use `-1` if required | Review scores are 1–5. `0` could be interpreted as a real score. |
| `review_comment_title`    | String    | `"No Title"`                                                         | Clearly indicates that the customer did not provide a title.     |
| `review_comment_message`  | String    | `"No Comment"`                                                       | Indicates that no review message was provided.                   |
| `review_creation_date`    | Timestamp | **Keep NULL** or use a clearly artificial date only if required      | A fake date can corrupt time-based analysis.                     |
| `review_answer_timestamp` | Timestamp | **Keep NULL**                                                        | NULL can legitimately mean the review was never answered.        |


**Not going forward with the above recommendation. Keeping the above field values as NULL, as No-information provided is some kind of information too.**

***

In [0]:
missing_values(orders_df, "orders")

Missing values in orders dataframe:
+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+
|order_id|customer_id|order_status|order_purchase_timestamp|order_approved_at|order_delivered_carrier_date|order_delivered_customer_date|order_estimated_delivery_date|
+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+
|       0|          0|           0|                       0|              160|                        1783|                         2965|                            0|
+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+



In [0]:
orders_df.filter(
                col("order_approved_at").isNull() |
                col("order_delivered_carrier_date").isNull() |
                col("order_delivered_customer_date").isNull() 
        ).show()

+--------------------+--------------------+------------+------------------------+-------------------+----------------------------+-----------------------------+-----------------------------+
|            order_id|         customer_id|order_status|order_purchase_timestamp|  order_approved_at|order_delivered_carrier_date|order_delivered_customer_date|order_estimated_delivery_date|
+--------------------+--------------------+------------+------------------------+-------------------+----------------------------+-----------------------------+-----------------------------+
|136cce7faa42fdb2c...|ed0271e0b7da060a3...|    invoiced|     2017-04-11 12:22:08|2017-04-13 13:25:17|                        NULL|                         NULL|          2017-05-09 00:00:00|
|ee64d42b8cf066f35...|caded193e8e47b836...|     shipped|     2018-06-04 16:44:48|2018-06-05 04:31:18|         2018-06-05 14:32:00|                         NULL|          2018-06-28 00:00:00|
|0760a852e4e9d89eb...|d2a79636084590b74...|  

For these order timestamp fields, I recommend being careful because NULL often represents a meaningful business status rather than bad data.

| Column                          | Meaning                              | Best handling                                                                                       |
| ------------------------------- | ------------------------------------ | --------------------------------------------------------------------------------------------------- |
| `order_approved_at`             | When the order was approved          | Usually keep `NULL`; optionally infer from `order_purchase_timestamp` only if business logic allows |
| `order_delivered_carrier_date`  | When the carrier received the order  | Keep `NULL` if the order was neither shipped <u>nor delivered or was cancelled</u>                                                           |
| `order_delivered_customer_date` | When the customer received the order | Keep `NULL` if the order was <u>not delivered or was cancelled</u>                                                         |


In [0]:
orders_df.select("order_status").distinct().show()

+------------+
|order_status|
+------------+
|     shipped|
|    canceled|
|    invoiced|
|     created|
|   delivered|
| unavailable|
|  processing|
|    approved|
+------------+



| Column                          | Replacement condition                              | Replace with                                                                                       |
| ------------------------------- | ------------------------------------ | --------------------------------------------------------------------------------------------------- |
| `order_approved_at`             | `order_status` = 'approved' OR 'shipped' OR 'delivered'          | Replace with (`order_purchase_timestamp` + 1 hour) |
| `order_delivered_carrier_date`  | `order_status` = 'shipped' OR 'delivered'  | Replace with (`order_approved_at` + 1 day)                                                           |
| `order_delivered_customer_date` | `order_status` = 'delivered' | Replace with `orderorder_estimated_delivery_date`   

In [0]:
orders_df.filter(col("order_status").isin("approved","shipped","delivered")).filter(col("order_approved_at").isNull()).show()

+--------------------+--------------------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+
|            order_id|         customer_id|order_status|order_purchase_timestamp|order_approved_at|order_delivered_carrier_date|order_delivered_customer_date|order_estimated_delivery_date|
+--------------------+--------------------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+
|e04abd8149ef81b95...|2127dc6603ac33544...|   delivered|     2017-02-18 14:40:00|             NULL|         2017-02-23 12:04:47|          2017-03-01 13:25:33|          2017-03-17 00:00:00|
|8a9adc69528e1001f...|4c1ccc74e00993733...|   delivered|     2017-02-18 12:45:31|             NULL|         2017-02-23 09:01:52|          2017-03-02 10:05:06|          2017-03-21 00:00:00|
|7013bcfc1c97fe719...|2941af76d38100e0f...|   delivered

In [0]:
from pyspark.sql.functions import col, coalesce, expr

cleaned_orders_df = (
                    orders_df
                        .withColumn(
                                "order_approved_at",
                                when(
                                    col("order_status").isin("approved","shipped","delivered"), 
                                    coalesce(
                                        col("order_approved_at"),
                                        col("order_purchase_timestamp").cast("timestamp") + expr("INTERVAL 1 HOUR")
                                    )
                                ).otherwise(
                                        col("order_approved_at")
                                )
                        )            
                    )



In [0]:
cleaned_orders_df.filter(col("order_status") == "canceled").show()

+--------------------+--------------------+------------+------------------------+-------------------+----------------------------+-----------------------------+-----------------------------+
|            order_id|         customer_id|order_status|order_purchase_timestamp|  order_approved_at|order_delivered_carrier_date|order_delivered_customer_date|order_estimated_delivery_date|
+--------------------+--------------------+------------+------------------------+-------------------+----------------------------+-----------------------------+-----------------------------+
|1b9ecfe83cdc25925...|6d6b50b66d79f8082...|    canceled|     2018-08-04 14:29:27|2018-08-07 04:10:26|                        NULL|                         NULL|          2018-08-14 00:00:00|
|714fb133a6730ab81...|e3fe72696c4713d64...|    canceled|     2018-01-26 21:34:08|2018-01-26 21:58:39|         2018-01-29 22:33:25|                         NULL|          2018-02-22 00:00:00|
|3a129877493c8189c...|0913cdce793684e52...|  

In [0]:
cleaned_orders_df.filter(col("order_status").isin("approved","shipped","delivered")).filter(col("order_approved_at").isNull()).show()

+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+
|order_id|customer_id|order_status|order_purchase_timestamp|order_approved_at|order_delivered_carrier_date|order_delivered_customer_date|order_estimated_delivery_date|
+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+
+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+



In [0]:
missing_values(cleaned_orders_df, "cleaned_orders")


Missing values in cleaned_orders dataframe:
+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+
|order_id|customer_id|order_status|order_purchase_timestamp|order_approved_at|order_delivered_carrier_date|order_delivered_customer_date|order_estimated_delivery_date|
+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+
|       0|          0|           0|                       0|              146|                        1783|                         2965|                            0|
+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+



In [0]:
cleaned_orders_df.filter(col("order_status").isin("shipped","delivered")).filter(col("order_delivered_carrier_date").isNull()).show()

+--------------------+--------------------+------------+------------------------+-------------------+----------------------------+-----------------------------+-----------------------------+
|            order_id|         customer_id|order_status|order_purchase_timestamp|  order_approved_at|order_delivered_carrier_date|order_delivered_customer_date|order_estimated_delivery_date|
+--------------------+--------------------+------------+------------------------+-------------------+----------------------------+-----------------------------+-----------------------------+
|2aa91108853cecb43...|afeb16c7f46396c0e...|   delivered|     2017-09-29 08:52:58|2017-09-29 09:07:16|                        NULL|          2017-11-20 19:44:47|          2017-11-14 00:00:00|
|2d858f451373b04fb...|e08caf668d499a6d6...|   delivered|     2017-05-25 23:22:43|2017-05-25 23:30:16|                        NULL|                         NULL|          2017-06-23 00:00:00|
+--------------------+--------------------+--

In [0]:
from pyspark.sql.functions import col, coalesce, expr

cleaned_orders_df = (
                    cleaned_orders_df
                        .withColumn(
                                "order_delivered_carrier_date",
                                when(
                                    col("order_status").isin("shipped","delivered"), 
                                    coalesce(
                                        col("order_delivered_carrier_date"),
                                        col("order_approved_at").cast("timestamp") + expr("INTERVAL 1 DAY")
                                    )
                                ).otherwise(
                                        col("order_delivered_carrier_date")
                                )
                        )            
                    )



In [0]:
cleaned_orders_df.filter(col("order_status").isin("shipped","delivered")).filter(col("order_delivered_carrier_date").isNull()).show()

+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+
|order_id|customer_id|order_status|order_purchase_timestamp|order_approved_at|order_delivered_carrier_date|order_delivered_customer_date|order_estimated_delivery_date|
+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+
+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+



In [0]:
missing_values(cleaned_orders_df, "cleaned_orders")


Missing values in cleaned_orders dataframe:
+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+
|order_id|customer_id|order_status|order_purchase_timestamp|order_approved_at|order_delivered_carrier_date|order_delivered_customer_date|order_estimated_delivery_date|
+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+
|       0|          0|           0|                       0|              146|                        1781|                         2965|                            0|
+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+



In [0]:
cleaned_orders_df.filter(col("order_status") == "delivered").filter(col("order_delivered_customer_date").isNull()).show()

+--------------------+--------------------+------------+------------------------+-------------------+----------------------------+-----------------------------+-----------------------------+
|            order_id|         customer_id|order_status|order_purchase_timestamp|  order_approved_at|order_delivered_carrier_date|order_delivered_customer_date|order_estimated_delivery_date|
+--------------------+--------------------+------------+------------------------+-------------------+----------------------------+-----------------------------+-----------------------------+
|2d1e2d5bf4dc7227b...|ec05a6d8558c6455f...|   delivered|     2017-11-28 17:44:07|2017-11-28 17:56:40|         2017-11-30 18:12:23|                         NULL|          2017-12-18 00:00:00|
|f5dd62b788049ad9f...|5e89028e024b381dc...|   delivered|     2018-06-20 06:58:43|2018-06-20 07:19:05|         2018-06-25 08:05:00|                         NULL|          2018-07-16 00:00:00|
|2ebdfc4f15f23b914...|29f0540231702fda0...|  

In [0]:
from pyspark.sql.functions import col, coalesce

cleaned_orders_df = (
                    cleaned_orders_df
                        .withColumn(
                                "order_delivered_customer_date",
                                when(
                                    col("order_status") == "delivered", 
                                    coalesce(
                                        col("order_delivered_customer_date"),
                                        col("order_estimated_delivery_date").cast("timestamp")
                                    )
                                ).otherwise(
                                        col("order_delivered_customer_date")
                                )
                        )            
                    )



In [0]:
cleaned_orders_df.filter(col("order_status") == "delivered").filter(col("order_delivered_customer_date").isNull()).show()

+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+
|order_id|customer_id|order_status|order_purchase_timestamp|order_approved_at|order_delivered_carrier_date|order_delivered_customer_date|order_estimated_delivery_date|
+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+
+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+



In [0]:
missing_values(cleaned_orders_df, "cleaned_orders")


Missing values in cleaned_orders dataframe:
+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+
|order_id|customer_id|order_status|order_purchase_timestamp|order_approved_at|order_delivered_carrier_date|order_delivered_customer_date|order_estimated_delivery_date|
+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+
|       0|          0|           0|                       0|              146|                        1781|                         2957|                            0|
+--------+-----------+------------+------------------------+-----------------+----------------------------+-----------------------------+-----------------------------+



***

In [0]:
missing_values(products_df, "products")

Missing values in products dataframe:
+----------+---------------------+-------------------+--------------------------+------------------+----------------+-----------------+-----------------+----------------+
|product_id|product_category_name|product_name_lenght|product_description_lenght|product_photos_qty|product_weight_g|product_length_cm|product_height_cm|product_width_cm|
+----------+---------------------+-------------------+--------------------------+------------------+----------------+-----------------+-----------------+----------------+
|         0|                  610|                610|                       610|               610|               2|                2|                2|               2|
+----------+---------------------+-------------------+--------------------------+------------------+----------------+-----------------+-----------------+----------------+



For the Olist products DataFrame, I would not treat all these NULLs the same way. Some are categorical, some are numerical, 
and some have only 2 missing values.

Here is a practical approach.

**Recommended treatment:**

| Column                       | Missing | Recommended treatment           | Reason                                                   |
| ---------------------------- | ------: | ------------------------------- | -------------------------------------------------------- |
| `product_category_name`      |     610 | `"Unknown"`                     | Missing category should not be invented                  |
| `product_name_lenght`        |     610 | `0` or derive from product name | If product name isn't available, 0 can represent no name |
| `product_description_lenght` |     610 | `0`                             | No description → length 0                                |
| `product_photos_qty`         |     610 | `0`                             | No photos → 0 photos                                     |
| `product_weight_g`           |       2 | **Median**                      | Only 2 missing; median is less sensitive to outliers     |
| `product_length_cm`          |       2 | **Median**                      | Same                                                     |
| `product_height_cm`          |       2 | **Median**                      | Same                                                     |
| `product_width_cm`           |       2 | **Median**                      | Same                                                     |


In [0]:
from pyspark.sql.functions import col, lit

# Fill categorical missing values in product_category
cleaned_products_df = products_df\
                            .fillna(value = "unknown", subset = ["product_category_name"])

# Fill numerical missing values in count/length fields
cleaned_products_df = cleaned_products_df.fillna({
                                "product_photos_qty": 0,
                                "product_name_lenght": 0,
                                "product_description_lenght": 0
                                })


In [0]:
missing_values(cleaned_products_df, "products")

Missing values in products dataframe:
+----------+---------------------+-------------------+--------------------------+------------------+----------------+-----------------+-----------------+----------------+
|product_id|product_category_name|product_name_lenght|product_description_lenght|product_photos_qty|product_weight_g|product_length_cm|product_height_cm|product_width_cm|
+----------+---------------------+-------------------+--------------------------+------------------+----------------+-----------------+-----------------+----------------+
|         0|                    0|                  0|                         0|                 0|               2|                2|                2|               2|
+----------+---------------------+-------------------+--------------------------+------------------+----------------+-----------------+-----------------+----------------+



***
**iii. Impute missing values (for numerical datatype columns)**

In [0]:
# For physical measurements missing values use "Median Imputation", 
# particularly because these are physical measurements and can have outliers...

from pyspark.ml.feature import Imputer

cols = [
    "product_weight_g",
    "product_length_cm",
    "product_height_cm",
    "product_width_cm"
]

imputer = Imputer(
                inputCols = cols,
                outputCols = cols
        ).setStrategy("mean") 


In [0]:
cleaned_products_df = imputer.fit(cleaned_products_df).transform(cleaned_products_df)

In [0]:
missing_values(cleaned_products_df, "products")

Missing values in products dataframe:
+----------+---------------------+-------------------+--------------------------+------------------+----------------+-----------------+-----------------+----------------+
|product_id|product_category_name|product_name_lenght|product_description_lenght|product_photos_qty|product_weight_g|product_length_cm|product_height_cm|product_width_cm|
+----------+---------------------+-------------------+--------------------------+------------------+----------------+-----------------+-----------------+----------------+
|         0|                    0|                  0|                         0|                 0|               0|                0|                0|               0|
+----------+---------------------+-------------------+--------------------------+------------------+----------------+-----------------+-----------------+----------------+



In [0]:
describe_df("products", cleaned_products_df)

DataFrame 'products' has 32951 rows and 9 columns.
--------------------------------------------------
Schema:
root
 |-- product_id: string (nullable = true)
 |-- product_category_name: string (nullable = false)
 |-- product_name_lenght: integer (nullable = false)
 |-- product_description_lenght: integer (nullable = false)
 |-- product_photos_qty: integer (nullable = false)
 |-- product_weight_g: integer (nullable = true)
 |-- product_length_cm: integer (nullable = true)
 |-- product_height_cm: integer (nullable = true)
 |-- product_width_cm: integer (nullable = true)

None
--------------------------------------------------
Showing top 10 dataframe: products's contents:


product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40,287,1,225,16,10,14
3aa071139cb16b67ca9e5dea641aaa2f,artes,44,276,1,1000,30,18,20
96bd76ec8810374ed1b65e291975717f,esporte_lazer,46,250,1,154,18,9,15
cef67bcfe19066a932b7673e239eb23d,bebes,27,261,1,371,26,4,26
9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37,402,4,625,20,17,13
41d3672d4792049fa1779bb35283ed13,instrumentos_musicais,60,745,1,200,38,5,11
732bd381ad09e530fe0a5f457d81becb,cool_stuff,56,1272,4,18350,70,24,44
2548af3e6e77a690cf3eb6368e9ab61e,moveis_decoracao,56,184,2,900,40,8,40
37cc742be07708b53a98702e77a21a02,eletrodomesticos,57,163,1,400,27,13,17
8c92109888e8cdf9d66dc7e463025574,brinquedos,36,1156,1,600,17,10,12


***
#### III. Removing Duplicates:-

In [0]:
##  Create a function to remove duplicates:

from pyspark.sql import DataFrame

def remove_duplicates(df: DataFrame, *cols: str) -> DataFrame:
    """
    Removes duplicate rows from a PySpark DataFrame based on a dynamic number of columns.
    
    :param df: The input PySpark DataFrame.
    :param cols: A dynamic sequence of column names to consider for deduplication.
                 If no columns are provided, it deduplicates across all columns.
    :return: A new DataFrame with duplicate rows removed.
    """
    # If no specific columns are passed, dropDuplicates() evaluates all columns
    if not cols:
        return df.dropDuplicates()
    
    # Otherwise, unpack the dynamic columns tuple into the subset parameter
    return df.dropDuplicates(subset=list(cols))

In [0]:
cleaned_customers_df = remove_duplicates(customers_df)
cleaned_geolocation_df = remove_duplicates(geolocation_df)
cleaned_order_items_df = remove_duplicates(order_items_df)
cleaned_order_payments_df = remove_duplicates(order_payments_df)
cleaned_order_reviews_df = remove_duplicates(cleaned_order_reviews_df)
cleaned_orders_df = remove_duplicates(cleaned_orders_df)
cleaned_products_df = remove_duplicates(cleaned_products_df)
cleaned_sellers_df = remove_duplicates(sellers_df)

In [0]:
describe_df("customers", cleaned_customers_df)

DataFrame 'customers' has 99441 rows and 5 columns.
--------------------------------------------------
Schema:
root
 |-- customer_id: string (nullable = true)
 |-- customer_unique_id: string (nullable = true)
 |-- customer_zip_code_prefix: integer (nullable = true)
 |-- customer_city: string (nullable = true)
 |-- customer_state: string (nullable = true)

None
--------------------------------------------------
Showing top 10 dataframe: customers's contents:


customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
eb03682e1118b3af176b229932bfa8be,9553802042571deac62d5789ed18a241,95765,bom principio,RS
55dbeedb7b2bf8d3059f319abe27202f,e44a59cd3178714eaf095face6112af5,23075,rio de janeiro,RJ
522a6e7c1f1f992ac81be6abefa46294,282ebddb3e85120b1652204758fc332b,60055,fortaleza,CE
277bdf9feec679fbb7319191a4a8daad,92d6a6df2cf2d6e2fa89dc1d28164c98,12503,guaratingueta,SP
a1674724305cdde2e900f8f50583f764,dd8c586651f9232e473ad86cf89c143f,28999,saquarema,RJ
b8f6bcb3f73c07a92dc1a5c28184e073,6bb3b4d34fb8f3c620d799442a568788,2443,sao paulo,SP
945c975073e57fe39a636afbf6689bcc,7ae8e196478928ab8223458a90681d26,13050,campinas,SP
35f68adc8ec170ab999117525306afc9,57e5587a5f1c3fcf65dff88c6d088cef,7197,guarulhos,SP
efef3994424e23ccbe4a144fd734e9af,b0b94079598f018f9e72e9698aefef33,12946,atibaia,SP
51965bd257c834f80a6dc936635a3ca4,7371189c8632ddaeb1e55bb5ed167adb,21511,rio de janeiro,RJ


In [0]:
describe_df("geolocation", cleaned_geolocation_df)

DataFrame 'geolocation' has 738332 rows and 5 columns.
--------------------------------------------------
Schema:
root
 |-- geolocation_zip_code_prefix: integer (nullable = true)
 |-- geolocation_lat: double (nullable = true)
 |-- geolocation_lng: double (nullable = true)
 |-- geolocation_city: string (nullable = true)
 |-- geolocation_state: string (nullable = true)

None
--------------------------------------------------
Showing top 10 dataframe: geolocation's contents:


geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
1048,-23.54651984318198,-46.6392350841075,sao paulo,SP
1026,-23.53925652899767,-46.633440525535235,sao paulo,SP
1041,-23.544084835131898,-46.63978569372155,são paulo,SP
1106,-23.531062064948905,-46.629505278925556,sao paulo,SP
1126,-23.52790740299499,-46.642218857204675,sao paulo,SP
1152,-23.52729286788048,-46.65889375200534,sao paulo,SP
1139,-23.51935174054843,-46.66240693563204,são paulo,SP
1236,-23.542477438253737,-46.66740182653356,sao paulo,SP
1050,-23.550449207051752,-46.64324288827981,sao paulo,SP
1026,-23.538895356079482,-46.63545499624909,sao paulo,SP


In [0]:
describe_df("order_items", cleaned_order_items_df)

DataFrame 'order_items' has 112650 rows and 7 columns.
--------------------------------------------------
Schema:
root
 |-- order_id: string (nullable = true)
 |-- order_item_id: integer (nullable = true)
 |-- product_id: string (nullable = true)
 |-- seller_id: string (nullable = true)
 |-- shipping_limit_date: timestamp (nullable = true)
 |-- price: double (nullable = true)
 |-- freight_value: double (nullable = true)

None
--------------------------------------------------
Showing top 10 dataframe: order_items's contents:


order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
01c4f4e08d9e8b7c5bd47e612285993f,1,422879e10f46682990de24d770e7f83d,1f50f920176fa81dab994f9023523100,2017-11-30T04:30:35Z,49.0,13.41
025884a2189930178d9ba3b1c6d0f9ba,1,a94036c0bb4dfce8d5762777c07b7be2,8160255418d5aaa7dbdc9f4c64ebda44,2018-03-26T09:48:16Z,124.9,27.6
03a3628434dd670dd5b9896337451c86,2,548ace38f22cc53db6f049c551d31397,79ed755314cfe6df0daef2c6cd3022cd,2017-07-26T19:35:21Z,179.9,17.02
0441b92c70bbdc1ae99a187e05d52621,1,d1dae3e210576b5698e7847df9c3e1a3,3c03b12bab54d8b37d79d914bfdb1aa0,2018-01-17T17:17:24Z,29.1,15.11
044496e055fca9c2f17e1309e77a48b6,1,b75bc6d95abd11862892008658b17f49,53e4c6e0f4312d4d2107a8c9cddf45cd,2017-09-22T08:10:07Z,131.0,25.23
045c7a03ba28c7d7b3c58e3897f587a7,1,4d68f0bb076a44142f1d1725ba3b3bc8,3a490ba60afa30ece0e7d50cfe74a4f0,2017-11-23T19:26:17Z,189.9,16.09
0489818f0b5863c095a7445ee35fd7f2,1,599e5320aef0c7dd9e3fed9b2f673d41,d03698c2efd04a549382afa6623e27fb,2018-03-21T21:10:29Z,81.4,23.5
04cc9ab9d21b11d7aca691cf7facaaa1,1,7c55ea4aea1acf1ce11440010f5aa298,955fee9216a65b617aa5c0531780ce60,2018-05-09T04:15:07Z,399.0,2.46
0505a29e827a4c609b3fd250c0313bef,1,282dcadbd8c61f52f06a8ae802b54ee6,f46490624488d3ff7ce78613913a7711,2018-06-18T04:31:40Z,99.9,42.76
0539642c2ba6dfd6c8ca8dbdb16d85d4,1,e5cac955339b48ea3b9773f034623e29,fa1c13f2614d7b5c4749cbc52fecda94,2018-03-01T15:31:28Z,189.9,12.83


In [0]:
describe_df("order_payments", cleaned_order_payments_df)

DataFrame 'order_payments' has 103886 rows and 5 columns.
--------------------------------------------------
Schema:
root
 |-- order_id: string (nullable = true)
 |-- payment_sequential: integer (nullable = true)
 |-- payment_type: string (nullable = true)
 |-- payment_installments: integer (nullable = true)
 |-- payment_value: double (nullable = true)

None
--------------------------------------------------
Showing top 10 dataframe: order_payments's contents:


order_id,payment_sequential,payment_type,payment_installments,payment_value
23f10f509600c30cf2852f9dbdf28fc3,1,credit_card,1,103.489998
2ccde1e4676c20b9444e2bcbed36334c,1,credit_card,8,134.559998
cf8ace9395d2ee06b4b513df8ba682b3,1,credit_card,1,52.3800011
1c859fbb6e7f238f3f6134628b845f3a,1,boleto,1,26.0900002
531752ff49fa411d865ecde8cce5c7ee,1,credit_card,2,57.7799988
9b92d4432c1758cd4215f3fdb1d623a3,1,credit_card,4,186.360001
1475652ad485c6469bd02393c70fdaa9,1,credit_card,2,81.7300034
bd7a44c3216be67318d6776fa920703d,1,credit_card,2,61.3699989
92af9bafd80bcd2636471ac01bc77f60,1,credit_card,8,204.110001
7c0b9b2385b65bd8c7aa778a843155e9,1,credit_card,2,151.600006


In [0]:
describe_df("order_reviews", cleaned_order_reviews_df)

DataFrame 'order_reviews' has 101895 rows and 7 columns.
--------------------------------------------------
Schema:
root
 |-- review_id: string (nullable = true)
 |-- order_id: string (nullable = true)
 |-- review_score: string (nullable = true)
 |-- review_comment_title: string (nullable = true)
 |-- review_comment_message: string (nullable = true)
 |-- review_creation_date: string (nullable = true)
 |-- review_answer_timestamp: string (nullable = true)

None
--------------------------------------------------
Showing top 10 dataframe: order_reviews's contents:


review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,null,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
f11d548dc1a7c713bff0b6d2ef0dff7c,0016dfedd97fc2950e388d2971d718c7,5,null,null,2017-05-23 00:00:00,2017-05-24 20:08:51
294845299f875a2abb84852d4c5be77b,59463b9c36d6a23c7c054114382f6d33,5,null,"Adorei! Pequeno, fácil de usar, não arranha apela e limpa a pele.. recomendo",2018-05-18 00:00:00,2018-05-21 11:33:57
5f222f80c589a6d83c18dc844bdb224c,1cc2afbca351a1423aeb95bd67821271,5,null,null,2018-04-19 00:00:00,2018-04-20 01:44:35
485d8767d19fdfb75652a75ffcc45955,fec1a2a1810b5d8cb301a4ddfebbd306,3,null,"Na propaganda está escrito Mixer elétrico e veio um portátil que eu tenho três que quebrou por isso eu queria um elétrico, vocês devem especificar melhor os produtos",2017-10-31 00:00:00,2017-11-01 21:54:14
d6d5cff1c0deb8e040dbb6d13db4aae7,be3522915c149522eb45df8854a71155,5,null,null,2018-05-25 00:00:00,2018-05-29 20:06:36
6a7eab0599692cb6bafd5782e2f26a01,d5cda33c87a415a696188f2f71799fe0,5,null,null,2018-07-06 00:00:00,2018-07-07 11:39:54
36d96c9b11e5041b2ac4940328443041,e7bf5b3305b73ad31a0306f433c8dd84,5,null,"Comunicação e respeito com o consumidor. É raro,comprar aqui sempre",2017-09-22 00:00:00,2017-09-23 11:17:43
b06f6882b1c0bc82ddf057b58b86fe50,2daee070f2042c8b7a8e9fdde778a31a,5,Muito bom,"Ótimo, excelente produto recomendo a todos",2018-07-12 00:00:00,2018-07-13 10:49:12
2286d28fbb1e4ffd4de21ac50544d1a7,3d0e84aa6695f156f958a00ca2a2745b,5,null,Bom demais,2018-02-15 00:00:00,2018-02-16 14:06:20


In [0]:
describe_df("orders", cleaned_orders_df)

DataFrame 'orders' has 99441 rows and 8 columns.
--------------------------------------------------
Schema:
root
 |-- order_id: string (nullable = true)
 |-- customer_id: string (nullable = true)
 |-- order_status: string (nullable = true)
 |-- order_purchase_timestamp: timestamp (nullable = true)
 |-- order_approved_at: timestamp (nullable = true)
 |-- order_delivered_carrier_date: timestamp (nullable = true)
 |-- order_delivered_customer_date: timestamp (nullable = true)
 |-- order_estimated_delivery_date: timestamp (nullable = true)

None
--------------------------------------------------
Showing top 10 dataframe: orders's contents:


order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
82daed31f80350ce51f518369a8d181c,d1240cdcdb9e6f08aa9a329bf1751c0e,delivered,2017-06-27T10:45:08Z,2017-06-27T11:04:48Z,2017-06-27T14:56:54Z,2017-06-29T16:45:36Z,2017-07-10T00:00:00Z
8adafb3466daa5395694d3a906ff9d40,b7919647bde69acc948baa47823d6c75,delivered,2017-01-25T15:47:27Z,2017-01-25T16:02:29Z,2017-02-07T08:09:23Z,2017-02-14T14:03:35Z,2017-03-23T00:00:00Z
4380bb6c7d8375926bbea2c8b514dc8c,11617a4c5fc72fccb4c2a53ea21c5a3c,delivered,2018-01-29T21:09:10Z,2018-01-29T21:32:17Z,2018-01-31T13:12:45Z,2018-02-14T21:43:34Z,2018-02-21T00:00:00Z
88c588c116fd06fb8e48dd83af88ec4a,8787c72a20002592238d7b80721ab729,delivered,2017-11-19T13:23:27Z,2017-11-19T13:35:26Z,2017-11-24T22:06:56Z,2017-12-22T17:44:27Z,2017-12-12T00:00:00Z
2d816b6a11ddd3843f307d1333321091,95303e4c9c23d67877e4b20afb04bfa8,delivered,2017-08-22T18:53:15Z,2017-08-22T19:06:03Z,2017-08-24T16:13:41Z,2017-08-25T20:12:31Z,2017-09-06T00:00:00Z
0cb25b24a27e27dd18273360995e99f4,b03b8bf457533ef245fd5fa2681732d0,delivered,2018-07-23T14:37:05Z,2018-07-23T16:00:30Z,2018-07-24T13:28:00Z,2018-07-25T23:36:35Z,2018-08-02T00:00:00Z
70462ebcf26934be8bd0327d2e6103d3,fb8b8298c590af3ee3169596c5ec2114,delivered,2018-08-20T22:06:10Z,2018-08-20T22:29:24Z,2018-08-21T13:49:00Z,2018-08-22T22:02:31Z,2018-08-28T00:00:00Z
4e21387c48145e5e07a1d552b1a69776,766a6dd9d0045e8462f001462976ff99,delivered,2018-02-06T16:17:58Z,2018-02-06T16:30:37Z,2018-02-07T16:28:28Z,2018-02-14T20:29:13Z,2018-03-02T00:00:00Z
ff8780cfb4535054cc5b56046b170cc0,f9835897d532524a6e9686847d4db4a7,delivered,2017-09-21T14:09:26Z,2017-09-22T05:34:34Z,2017-09-22T18:22:55Z,2017-10-01T14:33:15Z,2017-10-18T00:00:00Z
21094f6fd8017c8b7339040142abea71,faee0cb7b3e2c9fd9f57ec04cb70d51a,delivered,2018-07-24T18:10:05Z,2018-07-24T18:25:09Z,2018-07-30T14:17:00Z,2018-08-03T20:17:49Z,2018-08-24T00:00:00Z


In [0]:
describe_df("products", cleaned_products_df)

DataFrame 'products' has 32951 rows and 9 columns.
--------------------------------------------------
Schema:
root
 |-- product_id: string (nullable = true)
 |-- product_category_name: string (nullable = false)
 |-- product_name_lenght: integer (nullable = false)
 |-- product_description_lenght: integer (nullable = false)
 |-- product_photos_qty: integer (nullable = false)
 |-- product_weight_g: integer (nullable = true)
 |-- product_length_cm: integer (nullable = true)
 |-- product_height_cm: integer (nullable = true)
 |-- product_width_cm: integer (nullable = true)

None
--------------------------------------------------
Showing top 10 dataframe: products's contents:


product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
dc21247da0290904ffbbcd5953c76f95,papelaria,41,92,1,100,16,5,16
af16005fca813272caf59c432153949e,moveis_decoracao,55,1186,2,800,53,8,20
fe9184887637170f6da65623475d6a66,construcao_ferramentas_construcao,37,2457,1,600,26,17,17
3a78f64aac654298e4b9aff32fc21818,unknown,0,0,0,4300,30,10,48
80a96c144008b8446632f872d3b1a3ed,utilidades_domesticas,46,193,1,1025,25,23,25
5b706747080dcd34b3ebd803c25ccd65,relogios_presentes,52,654,4,340,47,13,13
be19e1611a6015a66266d199ff3ccce0,consoles_games,38,456,1,150,29,2,16
7e5fe07db814cff1e50a7561a10e37ef,cama_mesa_banho,54,672,2,250,16,10,16
a4c27cf9767fe2b589251c764b891eda,perfumaria,58,1061,2,368,22,15,18
5c0df984f68150bd43a1a59aaa83c13f,pet_shop,26,784,1,150,30,20,20


In [0]:
describe_df("sellers", cleaned_sellers_df)

DataFrame 'sellers' has 3095 rows and 4 columns.
--------------------------------------------------
Schema:
root
 |-- seller_id: string (nullable = true)
 |-- seller_zip_code_prefix: integer (nullable = true)
 |-- seller_city: string (nullable = true)
 |-- seller_state: string (nullable = true)

None
--------------------------------------------------
Showing top 10 dataframe: sellers's contents:


seller_id,seller_zip_code_prefix,seller_city,seller_state
f6122bc84774df1b372bdb3bb88ddb9f,6436,barueri,SP
8132b9bd16876e1b0f8808d43825dd48,88058,florianopolis,SC
a254c682cc01e119f83530446f1df9a9,16500,cafelandia,SP
cb9fb4ca75d7ba8437480e8dde64fe98,3551,sao paulo,SP
c5ebe6598748b0aeaa61cfb820478b92,95910,lajeado,RS
bd15ee794d5e640d9dd71b665b2ab15b,14078,ribeirao preto,SP
7994b065a7ffb14e71c6312cf87b9de2,29142,cariacica / es,ES
67883baaae6134ee81b271a542613728,1310,sao paulo,SP
f52c2422904463fdd7741f99045fecb6,9230,santo andre/sao paulo,SP
528bcf6680c36dddf07620bd35b33a6f,3178,sao paulo,SP


In [0]:
## Writing output files to destination folders:

storage_account = "ecommoliststorageaccount"
destination_path = f"abfss://olistdata@{storage_account}.dfs.core.windows.net/silver/01_cleaned_data"

# 1. Customer data

cleaned_customers_df.write.mode("overwrite").parquet(f"{destination_path}/cleaned_customers_data.parquet")

# 2. Geolocation data

cleaned_geolocation_df.write.mode("overwrite").parquet(f"{destination_path}/cleaned_geolocation_data.parquet")

# 3. Order items data

cleaned_order_items_df.write.mode("overwrite").parquet(f"{destination_path}/cleaned_order_items_data.parquet")

# 4. Order payments data

cleaned_order_payments_df.write.mode("overwrite").parquet(f"{destination_path}/cleaned_order_payments_data.parquet")

# 5. Order reviews data

cleaned_order_reviews_df.write.mode("overwrite").parquet(f"{destination_path}/cleaned_order_reviews_data.parquet")

# 6. Orders data

cleaned_orders_df.write.mode("overwrite").parquet(f"{destination_path}/cleaned_orders_data.parquet")

# 7. Products data

cleaned_products_df.write.mode("overwrite").parquet(f"{destination_path}/cleaned_products_data.parquet")

# 8. Sellers data

cleaned_sellers_df.write.mode("overwrite").parquet(f"{destination_path}/cleaned_sellers_data.parquet")


In [0]:
spark.stop()